WikiWatershed/model-my-watershed
# Prototype REST service

**Script to produce smaller GeoTiffs.**

A script has been added which produces compressed GeoTiffs of reduced color depth.  Useful for soil data, and possibly land-use, as well.  This script also makes the test soil rasters usable with the code (for testing).

Connects #546

**REST service.**

A spray-based REST service has been created.  The user POSTs a polygon stored as a GeoJSON file, and receives a count of land-use, soil-type pairs in the intersection of the query polygon and the test area.

Connects #545

The goals discussed in #544 can be done by sending a `GeometryCollection` instead of a `Polygon` or `MultiPolygon.  Those results can then be woven into a census of the type expected by the TR-55 code.

Connects #544

Big thanks to @echeipesh , @lossyrob , and @jwalgran .

**To Test**
   * Install spark (I used version 1.3.1)
   * Set the `SPARK_HOME` environment variable the name of the directory where spark is located
   * Set the `SPARK_VERSION` environment variable appropriately, in my case "1.3.1"
   * Apply this patch
   * Find the directory containing the three new tiff files that this patch creates; It should be `./src/geop/processing/src/test/data/` relative to the repository directory, set the MMW_DATA environment variable to the full name of that directory
   * Go to the `src/geop` directory (relative to the project root)
   * Type `sbt compile test` to run tests
   * Type `sbt 'project services' compile assembly` to combile the REST service code
   * Type `$SPARK_HOME/bin/spark-submit ./services/target/scala-2.10/mmw-services-assembly-0.0.1.jar` to run the service
   * Try to access the service on localhost:8107.  Anything other than POST should give you a 400.  Posting anything other than a valid GeoJSON polygon should give you a 400.
   * Sending a valid GeoJSON polygon should cause the service to return a JSON list of lists containing land-use X soil-type tuples and their counts (in the intersection of the query polygon and the small part of Delaware that we have data for)

Some example polygons follow.

This polygon does not intersect with Delaware, you should see no output (but no error).
```javascript
{
    "type": "Feature",
    "geometry": {
        "type": "Polygon",
        "coordinates": [
            [
                [31.137728, 29.975284],
                [31.128097, 29.972484],
                [31.130860, 29.976007],
                [31.134046, 29.979194],
                [31.137728, 29.975284]
            ]
        ]
    }
}
```

This polygon is a subset of the covered area.  You should see some output.
```javascript
{
    "type": "Feature",
    "geometry": {
        "type": "Polygon",
        "coordinates": [
            [
                [-75.72533952081164, 39.76],
                [-75.72533952081164, 39.8024743318554],
                [-75.58521226772196, 39.77],
                [-75.58521226772196, 39.7342308109474],
                [-75.72533952081164, 39.76]
            ]
        ]
    }
}
```

Here it is again as a multipolygon:
```javascript
{
    "type": "Feature",
    "geometry": {
        "type": "MultiPolygon",
        "coordinates": [
            [
                [
                    [-75.72533952081164, 39.76],
                    [-75.72533952081164, 39.8024743318554],
                    [-75.58521226772196, 39.77],
                    [-75.58521226772196, 39.7342308109474],
                    [-75.72533952081164, 39.76]
                ]
            ]
        ]
    }
}
```
Here it is again as a `GeometryCollection`.  You will get the same output as before, but wrapped in another list.
```javascript
{
    "type": "GeometryCollection",
    "geometries": [{
        "type": "Polygon",
        "coordinates": [
            [
                [-75.72533952081164, 39.76],
                [-75.72533952081164, 39.8024743318554],
                [-75.58521226772196, 39.77],
                [-75.58521226772196, 39.7342308109474],
                [-75.72533952081164, 39.76]
            ]
        ]
    }]
}
```

This is an invalid file, you should get a 400.
```javascript
{
    "type": "Feature",
    "properties": {
        "name": 6
    },
    "geometry": {
        "type": "Polygon",
        "coordinates": [
            [
                [-75.72533952081164, 39.76],
                [-75.72533952081164, 39.8024743318554],
                [-75.58521226772196, 39.77],
                [-75.58521226772196, 39.7342308109474],
                [-75.72533952081164, 39.76]
            ]
        ]
    }
```

You can POST files to the service by typing something like this `wget localhost:8080 --post=file=foo.json -O -`
